cli/popper/runner_host.py

```diff
@@ -36,225 +38,214 @@ def run(self, step):
+                    if logging:
+                        log.step_info(line)
+                    else:
+                        output.append(line)
+
+                p.wait()
+                ecode = p.poll()
+
+            log.debug(f'Code returned by process: {ecode}')
+
+        except SubprocessError as ex:
+            output = ""
+            ecode = ex.returncode
+            log.step_info(f"Command '{cmd[0]}' failed with: {ex}")
+        except Exception as ex:
+            output = ""
+            ecode = 1
+            log.step_info(f"Command raised non-SubprocessError error: {ex}")
+
+        return pid, ecode, '\n'.join(output)
 
 
 class DockerRunner(StepRunner):
-    """Runs steps in docker."""
-    d = None
+    """Runs steps in docker on the local machine."""
+    def __init__(self, init_docker_client=True, **kw):
+        super(DockerRunner, self).__init__(**kw)
 
-    # hold references to spawned containers
-    spawned_containers = []
+        self._spawned_containers = set()
+        self._d = None
 
-    def __init__(self, config):
-        super(DockerRunner, self).__init__(config)
+        if not init_docker_client:
+            return
 
         try:
-            DockerRunner.d = docker.from_env()
-            DockerRunner.d.version()
+            self._d = docker.from_env()
+            self._d.version()
         except Exception as e:
             log.debug(f'Docker error: {e}')
             log.fail(f'Unable to connect to the docker daemon.')
 
-        log.debug(f'Docker info: {pu.prettystr(DockerRunner.d.info())}')
+        log.debug(f'Docker info: {pu.prettystr(self._d.info())}')
 
     def __exit__(self, exc_type, exc_value, exc_traceback):
-        if DockerRunner.d:
-            DockerRunner.d.close()
-        DockerRunner.spawned_containers = []
-        return True
+        if self._d:
+            self._d.close()
+        self._spawned_containers = set()
 
     def run(self, step):
         """Execute the given step in docker."""
-        cid = pu.sanitized_name(step['name'], self.config.wid)
-
-        container = DockerRunner.find_container(cid)
+        cid = pu.sanitized_name(step['name'], self._config.wid)
 
-        if container and not self.config.reuse and not self.config.dry_run:
+        container = self._find_container(cid)
+        if container and not self._config.reuse and not self._config.dry_run:
             container.remove(force=True)
 
-        container = DockerRunner.create_container(cid, step, self.config)
+        container = self._create_container(cid, step)
 
         log.info(f'[{step["name"]}] docker start')
 
-        if self.config.dry_run:
+        if self._config.dry_run:
             return 0
 
-        DockerRunner.spawned_containers.append(container)
+        self._spawned_containers.add(container)
 
         container.start()
         cout = container.logs(stream=True)
         for line in cout:
-            log.step_info(pu.decode(line).strip('\n'))
```

Comment:
@ivotron  Without decoding and stripping the logs, byte strings are getting printed like this, 

```
b'hello\n'
```
Can we bring the decoding back ?

Reply:
I think we need to investigate why this is not decoded already. Based on the documentation, when `stream=True`, this is a generator of lines, so it should be returning text without new line characters

Reply:
@ivotron I just checked this by writing an independent script with docker py to spawn a container and stream logs. I found it prints out byte string by default.
```python3
import docker

client = docker.from_env()
c = client.containers.create('alpine:3.9', 'ls', detach=True)
c.start()
cout = c.logs(stream=True)
for line in cout:
	print(line)
```

gives output, 

```sh
b'bin\n'
b'dev\n'
b'etc\n'
b'home\n'
b'lib\n'
b'media\n'
b'mnt\n'
b'opt\n'
b'proc\n'
b'root\n'
b'run\n'
b'sbin\n'
b'srv\n'
b'sys\n'
b'tmp\n'
b'usr\n'
b'var\n'
```
So, looks like this is the default behaviour.

Reply:
thanks Jayjeet. The above reproduces the problem but doesn't explain why this is the default behavior. One option is to look at the docker-py code and see why when `stream=True` the library returns raw lines instead of decoded ones.

Reply:
Ok will do that

Reply:
Looking at the code of `docker-py`, I found this 
[this](https://github.com/docker/docker-py/blob/030af62dca982e62ed418314c857da1f90ce4282/docker/api/client.py#L393) function where if stream and tty is enabled, it looks like binary logs are streamed, but I am still not sure.

Reply:
aha! that's the answer. We need to do `tty=True` when we create the container

Reply:
I did `tty=True` but it then streams a sequence of bytes. I debugged the docker-py library all the combinations, looks like there's no way to get text output from `container.logs()`.

Reply:
I see. I misread the code, you're right. Since we don't support python2, we could do this directly:

```python
cout = container.logs(stream=True)
for raw_line in cout:
    log.step_info(raw_line.rstrip().decode('utf-8'))
```

we might need to the deal with this differently if `tty: True` is defined in the options. Something like docker-compose's [log printer](https://github.com/docker/compose/blob/abef11b2a6fa5bed4a86c3ebc022fb5e8898217a/compose/cli/log_printer.py), cause we also need to deal with the output of build and pull